File: Android/ItHappenedAndroid/app/src/main/java/ru/lod_misis/ithappened/Domain/Statistics/Facts/Models/Collections/Sequence.java
package ru.lod_misis.ithappened.Domain.Statistics.Facts.Models.Collections;

import java.util.ArrayList;
import java.util.List;

public class Sequence {

    private List<Double> Data;
    private double Sum;

    public Sequence(List<Double> data) {
        Data = data;
        for (Double d: data) {
            Sum += d;
        }
    }

    public double Sum() {
        return Sum;
    }

    public double Mean() {
        return Sum() / Data.size();
    }

    public int Length() {
        return Data.size();
    }

    public Double get(int i) {
        return Data.get(i);
    }

    public Sequence Mult(Sequence sequence) {
        if (sequence.Length() != Length()) {
            throw new IllegalArgumentException("Second sequence must have same length.");
        }
        List<Double> result = new ArrayList<>();
        for (int i = 0; i < Length(); i++) {
            result.add(get(i) * sequence.get(i));
        }
        return new Sequence(result);
    }

    public Sequence Pow(int degree) {
        List<Double> result = new ArrayList<>();
        for (int i = 0; i < Length(); i++) {
            result.add(Math.pow(get(i), degree));
        }
        return new Sequence(result);
    }

    public Sequence Slice(int start, int stop) {
        List<Double> result = new ArrayList<>();
        for (int i = start; i < stop; i++) {
            result.add(get(i));
        }
        return new Sequence(result);
    }

    public List<Double> ToList() {
        return Data;
    }
}

File: Backend/ItHappened/ARIMA/Models/Sequence.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ARIMA.Models
{
    internal class Sequence
    {
        private readonly List<double> data;
        private readonly double sum;

        public Sequence(List<double> data)
        {
            this.data = data;
            sum = 0;
            foreach (var elem in data)
            {
                sum += elem;
            }
        }

        public double this[int i] => data[i];
        public int Length() => data.Count;
        public double Sum() => sum;
        public double Mean() => sum / data.Count;

        public Sequence Sum(Sequence seq)
        {
            if (data.Count != seq.Length())
            {
                throw new ArgumentException("Dimensions of sequences should be same");
            }
            var result = new List<double>();
            for (var i = 0; i < seq.Length(); i++)
            {
                result.Add(data[i] + seq[i]);
            }
            return new Sequence(result);
        }

        public Sequence Sum(double constant)
        {
            var result = data.Select(x => x + constant).ToList();
            return new Sequence(result);
        }

        public Sequence Difference(Sequence seq)
        {
            if (data.Count != seq.Length())
            {
                throw new ArgumentException("Dimensions of sequences should be same");
            }
            var result = new List<double>();
            for (var i = 0; i < seq.Length(); i++)
            {
                result.Add(data[i] - seq[i]);
            }
            return new Sequence(result);
        }

        public Sequence Difference(double constant)
        {
            var result = data.Select(t => t - constant).ToList();
            return new Sequence(result);
        }

        public Sequence PowSequence(int degree)
        {
            var result = data.Select(t => Math.Pow(t, degree)).ToList();
            return new Sequence(result);
        }

        public Sequence Mult(Sequence seq)
        {
            if (data.Count != seq.Length())
            {
                throw new ArgumentException("Dimensions of sequences should be same");
            }
            var result = new List<double>();
            for (var i = 0; i < seq.Length(); i++)
            {
                result.Add(data[i] * seq[i]);
            }
            return new Sequence(result);
        }

        public Sequence Slice(int start, int end)
        {
            var result = new List<double>();
            for (var i = start; i < end; i++)
            {
                result.Add(data[i]);
            }
            return new Sequence(result);
        }
    }
}
